File: backend/app/schemas/company_report.py
"""Pydantic schemas for the per-company job status report.

Unlike ``daily_stats``, this report has no persisted table — every
request is computed live against ``jobs`` (see
``app.services.company_report``), so there is no ``Read`` model tied to
an ORM row here, just the response shape.
"""

from datetime import date, datetime
from uuid import UUID

from app.schemas.base import BaseSchema

REPORT_BUCKETS: tuple[str, ...] = (
    "rejected",
    "closed_completed",
    "scheduled_another_day",
    "canceled",
    "still_open",
)


class CompanyReportRow(BaseSchema):
    """One company's job-status breakdown for the requested date range."""

    company_id: UUID
    company_name: str
    rejected: int
    closed_completed: int
    scheduled_another_day: int
    canceled: int
    still_open: int
    total: int


class CompanyReportResponse(BaseSchema):
    start_date: date
    end_date: date
    items: list[CompanyReportRow]


class CompanyReportJobRow(BaseSchema):
    """One job behind a single company/bucket cell of the report — lets an
    operator confirm a job landed in the right bucket."""

    job_id: UUID
    dispatch_job_id: UUID | None
    bucket: str
    lifecycle_status: str
    matched_by: str
    """``"arrival"`` if ``first_message_at`` put this job in range, or
    ``"appointment"`` if it only qualifies because ``appt_at`` lands in
    range (it arrived on a different day) — set only when the caller
    passed ``include_scheduled_appts=True``."""
    first_message_at: datetime
    appt_at: datetime | None
    address: str | None
    customer_name: str | None
    customer_phone: str | None
    job_type: str | None
    message_preview: str | None


class CompanyReportJobsResponse(BaseSchema):
    """``bucket`` is one of ``REPORT_BUCKETS``, or ``"total"`` for the
    unfiltered "Total" column — every job for the company in range, across
    all buckets."""

    start_date: date
    end_date: date
    company_id: UUID
    company_name: str
    bucket: str
    items: list[CompanyReportJobRow]

File: frontend/src/types/company-report.ts
/**
 * Types for the live per-company job status report. Mirrors
 * ``backend/app/schemas/company_report.py``.
 *
 * Unlike daily stats, this is computed on every request — no
 * precomputed snapshot table — so "today" is always current.
 */

export interface CompanyReportRow {
  company_id: string;
  company_name: string;
  rejected: number;
  closed_completed: number;
  scheduled_another_day: number;
  canceled: number;
  still_open: number;
  total: number;
}

export interface CompanyReportResponse {
  start_date: string;
  end_date: string;
  items: CompanyReportRow[];
}

/** Bucket keys reported per company — matches ``REPORT_BUCKETS`` on the backend. */
export type CompanyReportBucket = Exclude<
  keyof CompanyReportRow,
  "company_id" | "company_name" | "total"
>;

export interface CompanyReportJobRow {
  job_id: string;
  dispatch_job_id: string | null;
  bucket: CompanyReportBucket;
  lifecycle_status: string;
  /**
   * "arrival" if first_message_at put this job in range; "appointment" if
   * it only qualifies because appt_at lands in range (arrived on a
   * different day) — only meaningful when include_scheduled_appts was on.
   */
  matched_by: "arrival" | "appointment";
  first_message_at: string;
  appt_at: string | null;
  address: string | null;
  customer_name: string | null;
  customer_phone: string | null;
  job_type: string | null;
  message_preview: string | null;
}

export interface CompanyReportJobsResponse {
  start_date: string;
  end_date: string;
  company_id: string;
  company_name: string;
  /** One of ``CompanyReportBucket``, or ``"total"`` for the unfiltered "Total" column. */
  bucket: CompanyReportBucket | "total";
  items: CompanyReportJobRow[];
}
